README.md
<h3 align="center"><img src="https://raw.githubusercontent.com/gencay/vscode-chatgpt/main/images/ai-logo.png" height="64"><br>An Unofficial VS Code - Chat GPT extension</h3>

<p align="center">
    <a href="https://marketplace.visualstudio.com/items?itemName=gencay.vscode-chatgpt" alt="Marketplace version">
        <img src="https://img.shields.io/visual-studio-marketplace/v/gencay.vscode-chatgpt?color=orange&label=VS%20Code" />
    </a>
    <a href="https://marketplace.visualstudio.com/items?itemName=gencay.vscode-chatgpt" alt="Marketplace download count">
        <img src="https://img.shields.io/visual-studio-marketplace/d/gencay.vscode-chatgpt?color=blueviolet&label=Downloads" />
    </a>
    <a href="https://github.com/gencay/vscode-chatgpt" alt="Github star count">
        <img src="https://img.shields.io/github/stars/gencay/vscode-chatgpt?color=blue&label=Github%20Stars" />
    </a>
</p>

## The most loved\* ChatGPT extension in VS Code open-sourced

> The project is built as a hobby project - With **no affiliation** to any organization
>
> The functionality fully relies on OpenAI services without any affiliation to them
>
> \*Downloaded by ~500,000 developers with more than 100 5 star ratings within 3 months of release

#### Try out this extension on marketplace: [ChatGPT - Genie AI](https://marketplace.visualstudio.com/items?itemName=genieai.chatgpt-vscode)

The extension is stripped out of Browser capabilities and published as an integration point for community to continue utilizing GPT models in Sidebar conversations or responses in editor, favorites, personalization and more.

## Forks and continuation of the original project

> Feel free to contribute here.

1. Continued by [ChatGPT - Genie AI](https://marketplace.visualstudio.com/items?itemName=genieai.chatgpt-vscode)
2. Forked by @Christopher-Hayes: [ChatGPT Reborn](https://marketplace.visualstudio.com/items?itemName=chris-hayes.chatgpt-reborn)

## Features

- ➕ Use GPT-4, GPT-3.5, GPT3 or Codex models using your OpenAI API Key
- 📃 Get streaming answers to your prompts in sidebar conversation window
- 🔥 Stop the responses to save your tokens.
- 📝 Create files or fix your code with one click or with keyboard shortcuts.
- ➡️ Export all your conversation history at once in Markdown format.
- Automatic partial code response detection. Continues and combines automatically, when response is cut off.
- Ad-hoc prompt prefixes for you to customize what you are asking ChatGPT
- Edit and resend a previous prompt
- Copy, insert or create new file from the code, ChatGPT is suggesting right into your editor.

## License

> If you are developing completely another extension and release it to the public, make sure you follow this.

This project is released under ISC License - See root License for details. Copyright notice and the respective permission notices must appear in all copies.

## How to build and run

- Clone the repository to your local machine
- On the root directory, run `yarn` command to install the dependencies listed in `package.json`
- Within VS Code - run the project by simply hitting `F5`

## How to install locally

- Install `vsce` if you don't have it on your machine (The Visual Studio Code Extension Manager)
  - `npm install --global vsce`
- Run `vsce package`
- Follow the <a href="https://code.visualstudio.com/docs/editor/extension-marketplace#_install-from-a-vsix">instructions</a> and install manually.


CHANGELOG.md
# Change Log

All notable changes to the "vscode-chatgpt" extension will be documented in this file.

## [3.9.6] - 2022-03-15

### 📢 ChatGPT Plus GPT-4 is now available via Browser autologin

- You can now use the GPT-4 via your ChatGPT Plus subscription on web using Browser autologin method
  > GPT-4 is only available on browser.

<img src="https://raw.githubusercontent.com/gencay/vscode-chatgpt/main/images/gpt4-model.png">

## [3.9.0] - 2022-03-02

### New features - Official ChatGPT API / GPT 3.5 Turbo support

- You can now use the following official ChatGPT APIs from OpenAI:
  - "OpenAI API Key - gpt-3.5-turbo"
  - "OpenAI API Key - gpt-3.5-turbo-0301"
- 'gpt-3.5-turbo' model is now the default model used.
- Make sure to update your model setting in your 'settings.json' if you previously used a different model.

<img src="https://raw.githubusercontent.com/gencay/vscode-chatgpt/main/images/gpt35-model.png">

## [3.8.0] - 2022-02-27

- Added stop generating responses feature.
  - You can now stop GPT to generate new tokens to save tokens if the response is too long.
- Improved Codex support.
- Added a new context menu item when `code-` model is selected: `ChatGPT-Codex: Generate code`
- Added a new setting to disable automatic scrolling of the responses as ChatGPT responds.

## [3.7.0] - 2022-02-25

#### ChatGPT Conversation History

- Use `Browser Auto-login` method to access all of your ChatGPT conversations.
- Either from the home page or the kebab menu, click `Show conversations`.
- Select a conversation to load.
- You can continue the conversation or move to another conversation easily.
- A timestamp for when the chat was created along with its title will be displayed on the top of the window.

## [3.6.0] - 2022-02-12

- Keybindings, new custom prompts and enable/disable prompts
- You can now use/customize keybindings for your prompts. (i.e. Use `cmd+k cmd+shift+1` to run `ChatGPT: Add Tests` on the selected code on MacOS)
- You can now enable/disable prompts that you wish to see in the editor's context menus.
- There are 2 new prompts that you can customize and enable if the currently available prompts are insufficient for your use case.

## [3.5.5] - 2022-02-05

- ChatGPT Plus support. If you have a Plus subscription, you can now use the \*-paid models for your prompts.

## [3.5.0] - 2022-02-04

- The pre-release Official ChatGPT model is available via OpenAI API Key method.

## [3.4.0] - 2022-01-20

- Streaming conversation is now available. You will see results as ChatGPT / Codex / GPT models are typing.

## [3.2.0] - 2022-01-15

- You can now use `Google` or `Microsoft` Authentication types during login flow to auto-fill the forms with the email/password you provide the extension with.

## [3.1.0] - 2022-01-12

- You can now have conversations with GPT3 using the Official OpenAI APIs.
- The GPT3/Codex dialogue support is similar to what ChatGPT offers. However, note that ChatGPT is more advanced in dialogues.

## [3.0.0] - 2022-01-06

- Added support for Official OpenAI APIs for GPT3 and Codex using API Keys.

## [2.6.0] - 2022-01-04

- Added proxy server. Addressing issue #32
- Added custom timeout. Addressing issue #33

## [2.3.0] - 2022-12-27

- You can now add custom prefix commands for the selected code. The extension will remember the prompt for subsequent queries. Addressing issue #28

## [2.2.0] - 2022-12-27

- You can now use any Chromium-based browser for autologin. Search for setting `chatgpt:chromiumpath` to override it. Check out the Readme for more.

## [2.1.0] - 2022-12-26

- Automatically login without needing to copy-paste session cookies is now live and enabled by default
- UX improvements: moved `Clear conversation` and `Export to MD` buttons to flyout.
- Improved error handling
- Misc. bug fixes

## [1.8.0] - 2022-12-14

- 🚀 [Experimental Feature] Automatically login without needing to copy-paste session cookies

## [1.7.0] - 2022-12-12

- 🚀 Automatically login without needing to copy-paste session cookies
- OpenAI servers are experiencing exceptionally high demand, which may cause intermittent errors in this extension.

## [1.6.0] - 2022-12-11

- 🆕 New vs-code commands to Clear and Export conversations

## [1.5.0] - 2022-12-11

- 🆕 Clear conversations and reset thread
- 🆕 Export all your conversation dialogue into a new tab with single click
- 🆕 Added a new introduction page that shows up before the conversation starts

## [1.4.3] - 2022-12-10

- 🚀 Better theming support in markdown

## [1.4.2] - 2022-12-10

- Update icons and SVGs used in conversation window.
- Show copy to clipboard buttons on hover for better UX.

## [1.4.0] - 2022-12-09

- Introduced edit and resend feature 🚀. Addressing #6
- Edit previous prompt in-place
- Better font-family support for code highlighter

## [1.3.0] - 2022-12-08

- Continuous conversations with ChatGPT with follow-ups. Addressing #7
- UX improvements

## [1.2.0] - 2022-12-07

- Automatically highlight code in conversation view
- UX improvements addressing #4 and #8

## [1.1.0] - 2022-12-07

- Adding a new configuration to opt-in to receive notifications when chatgpt responds.
  - This config is set to false by default to prevent notification spams.
  - If users are interested in it due to high-volume latencies, they can opt-in to get notifications when answers are ready.

## [1.0.0] - 2022-12-07

- Introducing new ways to interact with the suggested code
- Copy code to clipboard
- Insert code to text editor
- Create a new file with the ChatGPT suggested code

## [0.1.0] - 2022-12-6

- Initial release of the ChatGPT - VS Code extension.


chatgpt-5.1.1/index.js
// Adapted from https://github.com/transitive-bullshit/chatgpt-api

/**
 * 
 * MIT License

Copyright (c) 2023 Travis Fischer

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
 */

// src/chatgpt-api.ts
import Keyv from "keyv";
import pTimeout from "p-timeout";
import QuickLRU from "quick-lru";
import { v4 as uuidv4 } from "uuid";

// src/types.ts
var ChatGPTError = class extends Error {
};
var openai;
((openai2) => {
})(openai || (openai = {}));

// src/fetch.ts
var fetch = globalThis.fetch;

// src/fetch-sse.ts
import { createParser } from "eventsource-parser";

// src/stream-async-iterable.ts
async function* streamAsyncIterable(stream) {
  const reader = stream.getReader();
  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) {
        return;
      }
      yield value;
    }
  } finally {
    reader.releaseLock();
  }
}

// src/fetch-sse.ts
async function fetchSSE(url, options, fetch2 = fetch) {
  const { onMessage, ...fetchOptions } = options;
  const res = await fetch2(url, fetchOptions);
  if (!res.ok) {
    let reason;
    try {
      reason = await res.text();
    } catch (err) {
      reason = res.statusText;
    }
    const msg = `ChatGPT error ${res.status}: ${reason}`;
    const error = new ChatGPTError(msg, { cause: res });
    error.statusCode = res.status;
    error.statusText = res.statusText;
    error.reason = reason;
    throw error;
  }
  const parser = createParser((event) => {
    if (event.type === "event") {
      onMessage(event.data);
    }
  });
  if (!res.body.getReader) {
    const body = res.body;
    if (!body.on || !body.read) {
      throw new ChatGPTError('unsupported "fetch" implementation');
    }
    body.on("readable", () => {
      let chunk;
      while (null !== (chunk = body.read())) {
        parser.feed(chunk.toString());
      }
    });
  } else {
    for await (const chunk of streamAsyncIterable(res.body)) {
      const str = new TextDecoder().decode(chunk);
      parser.feed(str);
    }
  }
}

// src/chatgpt-api.ts
var CHATGPT_MODEL = "gpt-3.5-turbo";
var USER_LABEL_DEFAULT = "User";
var ASSISTANT_LABEL_DEFAULT = "ChatGPT";
var ChatGPTAPI = class {
  /**
   * Creates a new client wrapper around OpenAI's chat completion API, mimicing the official ChatGPT webapp's functionality as closely as possible.
   *
   * @param apiKey - OpenAI API key (required).
   * @param apiBaseUrl - Optional override for the OpenAI API base URL.
   * @param debug - Optional enables logging debugging info to stdout.
   * @param completionParams - Param overrides to send to the [OpenAI chat completion API](https://platform.openai.com/docs/api-reference/chat/create). Options like `temperature` and `presence_penalty` can be tweaked to change the personality of the assistant.
   * @param maxModelTokens - Optional override for the maximum number of tokens allowed by the model's context. Defaults to 4096.
   * @param maxResponseTokens - Optional override for the minimum number of tokens allowed for the model's response. Defaults to 1000.
   * @param messageStore - Optional [Keyv](https://github.com/jaredwray/keyv) store to persist chat messages to. If not provided, messages will be lost when the process exits.
   * @param getMessageById - Optional function to retrieve a message by its ID. If not provided, the default implementation will be used (using an in-memory `messageStore`).
   * @param upsertMessage - Optional function to insert or update a message. If not provided, the default implementation will be used (using an in-memory `messageStore`).
   * @param organization - Optional organization string for openai calls.
   * @param fetch - Optional override for the `fetch` implementation to use. Defaults to the global `fetch` function.
   */
  constructor(opts) {
    const {
      apiKey,
      apiBaseUrl = "https://api.openai.com",
      organization,
      debug = false,
      messageStore,
      completionParams,
      systemMessage,
      maxModelTokens = 4e3,
      maxResponseTokens = 1e3,
      getMessageById,
      upsertMessage,
      fetch: fetch2 = fetch
    } = opts;
    this._apiKey = apiKey;
    this._apiBaseUrl = apiBaseUrl;
    this._organization = organization;
    this._debug = !!debug;
    this._fetch = fetch2;
    this._completionParams = {
      model: CHATGPT_MODEL,
      temperature: 0.8,
      top_p: 1,
      presence_penalty: 1,
      ...completionParams
    };
    this._systemMessage = systemMessage;
    if (this._systemMessage === void 0) {
      const currentDate = (/* @__PURE__ */ new Date()).toISOString().split("T")[0];
      this._systemMessage = `You are ChatGPT, a large language model trained by OpenAI. Answer as concisely as possible.
Knowledge cutoff: 2021-09-01
Current date: ${currentDate}`;
    }
    this._maxModelTokens = maxModelTokens;
    this._maxResponseTokens = maxResponseTokens;
    this._getMessageById = getMessageById ?? this._defaultGetMessageById;
    this._upsertMessage = upsertMessage ?? this._defaultUpsertMessage;
    if (messageStore) {
      this._messageStore = messageStore;
    } else {
      this._messageStore = new Keyv({
        store: new QuickLRU({ maxSize: 1e4 })
      });
    }
    if (!this._apiKey) {
      throw new Error("OpenAI missing required apiKey");
    }
    if (!this._fetch) {
      throw new Error("Invalid environment; fetch is not defined");
    }
    if (typeof this._fetch !== "function") {
      throw new Error('Invalid "fetch" is not a function');
    }
  }
  /**
   * Sends a message to the OpenAI chat completions endpoint, waits for the response
   * to resolve, and returns the response.
   *
   * If you want your response to have historical context, you must provide a valid `parentMessageId`.
   *
   * If you want to receive a stream of partial responses, use `opts.onProgress`.
   *
   * Set `debug: true` in the `ChatGPTAPI` constructor to log more info on the full prompt sent to the OpenAI chat completions API. You can override the `systemMessage` in `opts` to customize the assistant's instructions.
   *
   * @param message - The prompt message to send
   * @param opts.parentMessageId - Optional ID of the previous message in the conversation (defaults to `undefined`)
   * @param opts.messageId - Optional ID of the message to send (defaults to a random UUID)
   * @param opts.systemMessage - Optional override for the chat "system message" which acts as instructions to the model (defaults to the ChatGPT system message)
   * @param opts.timeoutMs - Optional timeout in milliseconds (defaults to no timeout)
   * @param opts.onProgress - Optional callback which will be invoked every time the partial response is updated
   * @param opts.abortSignal - Optional callback used to abort the underlying `fetch` call using an [AbortController](https://developer.mozilla.org/en-US/docs/Web/API/AbortController)
   * @param completionParams - Optional overrides to send to the [OpenAI chat completion API](https://platform.openai.com/docs/api-reference/chat/create). Options like `temperature` and `presence_penalty` can be tweaked to change the personality of the assistant.
   *
   * @returns The response from ChatGPT
   */
  async sendMessage(text, opts = {}) {
    const {
      parentMessageId,
      messageId = uuidv4(),
      timeoutMs,
      onProgress,
      stream = onProgress ? true : false,
      completionParams
    } = opts;
    let { abortSignal } = opts;
    let abortController = null;
    if (timeoutMs && !abortSignal) {
      abortController = new AbortController();
      abortSignal = abortController.signal;
    }
    const message = {
      role: "user",
      id: messageId,
      parentMessageId,
      text
    };
    await this._upsertMessage(message);
    const { messages } = await this._buildMessages(text, opts);
    const result = {
      role: "assistant",
      id: uuidv4(),
      parentMessageId: messageId,
      text: ""
    };
    const responseP = new Promise(
      async (resolve, reject) => {
        var _a, _b;
        const url = `${this._apiBaseUrl}/v1/chat/completions`;
        const headers = {
          "Content-Type": "application/json",
          Authorization: `Bearer ${this._apiKey}`
        };
        if (this._organization) {
          headers["OpenAI-Organization"] = this._organization;
        }
        const body = {
          ...this._completionParams,
          ...completionParams,
          messages,
          stream
        };
        if (stream) {
          fetchSSE(
            url,
            {
              method: "POST",
              headers,
              body: JSON.stringify(body),
              signal: abortSignal,
              onMessage: (data) => {
                var _a2;
                if (data === "[DONE]") {
                  result.text = result.text.trim();
                  return resolve(result);
                }
                try {
                  const response = JSON.parse(data);
                  if (response.id) {
                    result.id = response.id;
                  }
                  if ((_a2 = response == null ? void 0 : response.choices) == null ? void 0 : _a2.length) {
                    const delta = response.choices[0].delta;
                    result.delta = delta.content;
                    if (delta == null ? void 0 : delta.content)
                      result.text += delta.content;
                    result.detail = response;
                    if (delta.role) {
                      result.role = delta.role;
                    }
                    onProgress == null ? void 0 : onProgress(result);
                  }
                } catch (err) {
                  console.warn("OpenAI stream SEE event unexpected error", err);
                  return reject(err);
                }
              }
            },
            this._fetch
          ).catch(reject);
        } else {
          try {
            const res = await this._fetch(url, {
              method: "POST",
              headers,
              body: JSON.stringify(body),
              signal: abortSignal
            });
            if (!res.ok) {
              const reason = await res.text();
              const msg = `OpenAI error ${res.status || res.statusText}: ${reason}`;
              const error = new ChatGPTError(msg, { cause: res });
              error.statusCode = res.status;
              error.statusText = res.statusText;
              return reject(error);
            }
            const response = await res.json();
            if (this._debug) {
              console.log(response);
            }
            if (response == null ? void 0 : response.id) {
              result.id = response.id;
            }
            if ((_a = response == null ? void 0 : response.choices) == null ? void 0 : _a.length) {
              const message2 = response.choices[0].message;
              result.text = message2.content;
              if (message2.role) {
                result.role = message2.role;
              }
            } else {
              const res2 = response;
              return reject(
                new Error(
                  `OpenAI error: ${((_b = res2 == null ? void 0 : res2.detail) == null ? void 0 : _b.message) || (res2 == null ? void 0 : res2.detail) || "unknown"}`
                )
              );
            }
            result.detail = response;
            return resolve(result);
          } catch (err) {
            return reject(err);
          }
        }
      }
    ).then((message2) => {
      return this._upsertMessage(message2).then(() => message2);
    });
    if (timeoutMs) {
      if (abortController) {
        ;
        responseP.cancel = () => {
          abortController.abort();
        };
      }
      return pTimeout(responseP, {
        milliseconds: timeoutMs,
        message: "OpenAI timed out waiting for response"
      });
    } else {
      return responseP;
    }
  }
  get apiKey() {
    return this._apiKey;
  }
  set apiKey(apiKey) {
    this._apiKey = apiKey;
  }
  async _buildMessages(text, opts) {
    const { systemMessage = this._systemMessage } = opts;
    let { parentMessageId } = opts;
    const userLabel = USER_LABEL_DEFAULT;
    const assistantLabel = ASSISTANT_LABEL_DEFAULT;
    let messages = [];
    if (systemMessage) {
      messages.push({
        role: "system",
        content: systemMessage
      });
    }
    const systemMessageOffset = messages.length;
    let nextMessages = text ? messages.concat([
      {
        role: "user",
        content: text,
        name: opts.name
      }
    ]) : messages;
    do {
      const prompt = nextMessages.reduce((prompt2, message) => {
        switch (message.role) {
          case "system":
            return prompt2.concat([`Instructions:
${message.content}`]);
          case "user":
            return prompt2.concat([`${userLabel}:
${message.content}`]);
          default:
            return prompt2.concat([`${assistantLabel}:
${message.content}`]);
        }
      }, []).join("\n\n");
      messages = nextMessages;
      if (!parentMessageId) {
        break;
      }
      const parentMessage = await this._getMessageById(parentMessageId);
      if (!parentMessage) {
        break;
      }
      const parentMessageRole = parentMessage.role || "user";
      nextMessages = nextMessages.slice(0, systemMessageOffset).concat([
        {
          role: parentMessageRole,
          content: parentMessage.text,
          name: parentMessage.name
        },
        ...nextMessages.slice(systemMessageOffset)
      ]);
      parentMessageId = parentMessage.parentMessageId;
    } while (true);
    return { messages };
  }
  // protected get _isCodexModel() {
  //   return this._completionParams.model.startsWith('code-')
  // }
  async _defaultGetMessageById(id) {
    const res = await this._messageStore.get(id);
    return res;
  }
  async _defaultUpsertMessage(message) {
    await this._messageStore.set(message.id, message);
  }
};
export {
  ChatGPTAPI,
  ChatGPTError,
  openai
};
//# sourceMappingURL=index.js.map
